Add debugging a ci dump documentation template and script to fill info and include it on azdo test tab

safern opened the pull request.

safern committed:
Add debugging a ci dump documentation template and script to generate it on helix

danmosemsft commented:
@mikem8361

safern commented:
I want to do this as an initial step. The goal is that the .md file appears side by side with the dump on the Azure DevOps test tab. However, once we have a symbol support for helix: https://github.com/dotnet/arcade/issues/6480 -- we could very much instead generate a python script on the fly, that people would download and would setup everything for them, basically run the steps in the .md file automatically. Thoughts on that?

mikem8361 commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,122 @@
+
+### Debug with WinDbg
+
+1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).
```
With the "internal" windbg sos.dll should be automatically be loaded from the extension gallery. windbg/cdb can be installed from \\dbg\privates\rc or \\dbg\privates\latest so installing dotnet-sos isn't necessary. It is a lot more manual process too, FYI, `dotnet tool install -g dotnet-sos`,  `dotnet-sos install` which displays the command too load this sos.dll like'.load C:\Users\mikem\.dotnet\sos\sos.dll'.

Linux dumps should also work with the latest windbg.

safern replied:
I think we shouldn't add internal instructions here since this can also be done by contributors... but I'm fine if we add a note for that.

mikem8361 replied:
Yes, I didn't think about extern contributors. They don't have access to this version of windbg and the extension gallery. So I guess you are on the right track with the more explicit instructions about installing dotnet-sos and explicitly loading sos. Just remember that in the internal windbg will already be already loaded with what should be the same version as dotnet-sos install.

mikem8361 commented on eng/testing/debug-dump-template.md:
````diff
@@ -0,0 +1,122 @@
+
+```script
+!setclrpath <path to core root or testhost where coreclr is>
+!sethostruntime '<path to your local dotnet runtime or testhost>'
````
If devs have .NET Core installed sos will find it.  !sethostruntime is only necessary if it isn't or installed in a non-standard path.

mikem8361 commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,122 @@
+
+### Debug with WinDbg
+
+1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).
```
Same thing as above.

mikem8361 commented on eng/testing/debug-dump-template.md:
````diff
@@ -0,0 +1,122 @@
+3. Then run the following commands:
+
+```script
+.load '<path to sos (can get it by running `dotnet-sos install`)>'
````
If you go with the manual load route, this line should also be in the above script for Windows dumps.

hoyosjs replied:
And make sure to unload the one that might be bundled first. This is only necessary if the version of SOS in the gallery doesn't have the needed commands and bug fixes yet. This is probably not all too common.

safern replied:
When opening the Linux dump it didn't automatically load the sos plugin for me, but when opening a Windows dump it did... do we know why? I'll have to try this again.

hoyosjs replied:
On linux, did you run `dotnet sos install`?

mikem8361 replied:
I think he means linux dumps under windbg and I think there is a bug that I need to follow up on that sos doesn't get loaded for linux dumps (triggered by the libcoreclr.so module). The command `.update sos` should load it. I need to follow up with the windbg about why it isn't automatically being loaded.

mikem8361 replied:
I have version 10.0.20212.1001 and it loads sos.dll for my linux dumps now. Maybe you need a newer windbg version.

safern replied:
> I have version 10.0.20212.1001 and it loads sos.dll for my linux dumps now. Maybe you need a newer windbg version.

Ah that's it. I have a quite old version. Does it make sense mentioning `.update sos` command just in case?

mikem8361 replied:
You can mention it, but it would be better to make everybody has a new version.

safern replied:
Sounds good.

mikem8361 commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,122 @@
+setclrpath <path to core root or testhost where coreclr is>
+sethostruntime '<path to your local dotnet runtime or testhost where coreclr is>'
+setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
+# On linux (native symbols are loaded separately on lldb using `target symbols add <symbol file>`)
```
The `loadsymbols` command should also try to find/download/load the native module's symbols.

hoyosjs replied:
This is new to me. Thanks Mike. I usually had to load coreclr's separately for some reason.

mikem8361 commented:
/cc @hoyosjs

hoyosjs commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -210,6 +210,22 @@
+        $(HelixPostCommands);
+        %HELIX_PYTHONPATH% %HELIX_CORRELATION_PAYLOAD%\gen-debug-dump-docs.py -buildid $(BUILD_ID) -workitem %HELIX_WORKITEM_ID% -jobid %HELIX_CORRELATION_ID% -outdir %HELIX_WORKITEM_UPLOAD_ROOT%
+      </HelixPostCommands>
+      <HelixPostCommands Condition="'$(TargetsWindows)' != 'true'">
```
One very big caveat here is macOS. It's only supported to debug them in dotnet-dump if it's a `createdump` dump. If it's a system dump, then only lldb works.

safern replied:
I see. I'll mention it in the doc at the end.

MattGal commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -210,6 +210,22 @@
+    </PropertyGroup>
+
+    <ItemGroup Condition="'$(RuntimeFlavor)' == 'CoreCLR'">
+      <HelixCorrelationPayload Include="$(RepositoryEngineeringDir)testing\debug-dump-template.md" />
```
Payloads must be archives, probably easiest for this is to just use [the ZipDirectory task](https://docs.microsoft.com/en-us/visualstudio/msbuild/zipdirectory-task?view=vs-2019).

safern committed:
PR Feedback and fix helix workitems

safern committed:
Fix python script and -buildid argument

safern committed:
Upload doc only if dumps were found

safern committed:
Pass down templatedir

safern commented:
This is now working:

![image](https://user-images.githubusercontent.com/22899328/97644778-76324a80-1a08-11eb-93e3-92c79ebf8df7.png)

https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-229762be8cb749d196/Microsoft.CSharp.Tests/debug-dump.md?sv=2019-07-07&se=2020-11-18T23%3A31%3A30Z&sr=c&sp=rl&sig=YzLMjEzJP8W3zhGSp5m2JVzphr%2FgNVm0xRO%2FnV4wNJk%3D

cc: @ViktorHofer @danmosemsft

hoyosjs commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,126 @@
+
+### Debug with WinDbg
+
+1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).
```
The user still needs to run `dotnet sos install`. Perhaps also worth noting - this has an architecture flag to install different versions of the plugin for scenarios like arm on arm64 and such.

safern replied:
Yeah good suggestion.

hoyosjs commented on eng/testing/debug-dump-template.md:
````diff
@@ -0,0 +1,126 @@
+3. Then run the following commands:
+
+```script
+setclrpath <you should see `Load path for DAC/DBI: '<none>'`>
````
There's no distinction here between the two `setclrpath`. The first one is to verify a wrong DAC hasn't been loaded. That one has no parameters. The second one should have a path to the folder containing coreclr as the parameter.

hoyosjs commented on src/libraries/sendtohelix.proj:
```diff
@@ -144,7 +144,22 @@
     <_RuntimeInputs Condition=" '$(Scenarios)' != '' and '$(TargetsWindows)' != 'true' " Include="$(TestHostRootPath)**/*.sh" />
   </ItemGroup>
 
+  <Target Name="IncludeDumpDocsInTesthost"
```
Should this also be added for CoreCLR?

safern replied:
CoreCLR uses another sendtohelix project. I can hook it up there. However, the CoreCLR payloads don't contain symbols.. so those instructions are going to be a little bit different.

safern replied:
Will do this as part of a follow up PR.

hoyosjs commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,126 @@
+
+## MacOS dumps
+
+MacOS instructions are the same as [Linux](#linux-dumps-on-linux); however MacOS has some caveats.
```
```suggestion
Instructions for debugging dumps on MacOS the same as [Linux](#linux-dumps-on-linux); however there are a couple of caveats.
```

jkotas commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,126 @@
+
+Once you have those assets, you will need to extract the testhost or core root. Then extract the workitem assets into the same location where coreclr binary is.
+
+## Windows dump on windows
```
```suggestion
## Windows dump on Windows
```

jkotas commented on eng/testing/debug-dump-template.md:
````diff
@@ -0,0 +1,126 @@
+setsymbolserver -directory <directory with symbols (for library tests these are in testhost dir)>
+```
+
+## Linux dumps on windows
````
```suggestion
## Linux dumps on Windows
```

danmosemsft commented:
Awesome!!! 😁

hoyosjs commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,126 @@
+
+MacOS instructions are the same as [Linux](#linux-dumps-on-linux); however MacOS has some caveats.
+
+1. It's only supported to debug them in `dotnet-dump` if it's a `createdump` dump.
```
```suggestion
1. It's only supported to debug them in `dotnet-dump` if it's a runtime generated dump. This includes hang dumps and dumps generated by `createdump` and `dotnet-dump` and the runtime itself. 
```

hoyosjs reviewed (commented):
Looking good. Just a couple of comments and nits

danmosemsft commented:
Since we don't control the text on this page presumably (eg "Uploaded files") .. would naming this something like "how-to-debug-dumps.md" make it more likely someone will realize they can click on it for help?

To me "debug-dump.md" at a glance looked like a dump file itself, or some metadata about it, not "help"

stephentoub commented on eng/testing/debug-dump-template.md:
````diff
@@ -0,0 +1,126 @@
+You can just achieve this by running:
+
+```script
+runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o <out-dir>
````
Can you outline from where someone gets the relevant jobid and workitem?

safern replied:
This will be filled in by the python script. So the one that will show up by the dump will contain that info there already, that's why I didn't put that info. I was going to put it but then when I so the produced file it didn't make sense. Thoughts on that?

stephentoub replied:
You're saying I literally write `-j %JOBID% -w %WORKITEM%` rather than supplying those values myself?  Why do I even need to write that then, i.e. why is the command not just `runfo get-helix-payload -o <out-dir>`?

safern replied:
No not really. What I'm saying is that this is just a template that we feed into the python script that I'm adding as part of this PR and the python script will fill in this information and produce a .md file with all the information in it. For example, this .md file was generated from a crash I forced: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-229762be8cb749d196/Microsoft.CSharp.Tests/debug-dump.md?sv=2019-07-07&se=2020-11-18T23%3A31%3A30Z&sr=c&sp=rl&sig=YzLMjEzJP8W3zhGSp5m2JVzphr%2FgNVm0xRO%2FnV4wNJk%3D

You can see that the command it has is: 
```
runfo get-helix-payload -j 229762be-8cb7-49d1-960a-83e041494e15 -w a760fa7c-e088-41ca-b565-640db87df271 -o <out-dir>
```

So what I was trying to say, is that if you just read the template checked in into the repo, it makes sense to add the information from where to get the jobid... however the purpose of this template is to produce a .md file that is side by side with the dump in the helix logs container, so in the final .md file it will be confusing to have a section to explain "how to get the helix jobid" when you don't even need it since the jobid is already embedded into the .md file you're reading. 

Does that makes it more clear?

stephentoub replied:
Sorry, I'm still not understanding.  The directions say all I have to do is:
```
dotnet tool install --global runfo
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o outputdir
```
How does it know what job I care about / what to download?  So is there a step missing from the instructions?

hoyosjs replied:
This is a template. It gets populated before it the upload to the AzDO tab.

stephentoub replied:
Populated with what?  I haven't told it what I care about.

I'm confused.  I want to debug an issue that occurred in CI (which is what this doc is helping me with).  What exactly do I type at the command line?  Let's say my PR for dotnet/runtime PR 12345 failed.  What exactly do I do?

safern replied:
It will be populated with the actual helix job id and workitem id for the crash that you hit in CI.

We populate it here: https://github.com/dotnet/runtime/pull/43937/files#diff-fd8bb0da491f288f6e268f835a49b14126ba3f1a15ab6971ddc76dd480b81b11R95

We pass down that information here: https://github.com/dotnet/runtime/pull/43937/files#diff-1d68a522b60c1ab69fb537ce06f5d7e9a176fa4f75e16425fc164d4b738d7678R216

> What exactly do I type at the command line? Let's say my PR for dotnet/runtime PR 12345 failed. What exactly do I do?

You will go to the azure devops test tab which will show the failure and then in the helix attachments you will see the dump and a MarkDown file. That MD file, will be this template but with the info that you need embedded in there so that you don't have to go and get the helix job yourself. It will look like this:

![image](https://user-images.githubusercontent.com/22899328/97765068-f79fe080-1acd-11eb-9cae-82f91418f382.png)

So the steps you'll follow is download the dump and debug-dump.md file (which I will rename to how-to-debug-dump.md) and just follow the instructions in the MD file, no need to get more information. This is trying to minimize the steps people have to do and info they need to gather whenever they see a dump in their test results. The final MD file will look like [this one](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-229762be8cb749d196/Microsoft.CSharp.Tests/debug-dump.md?sv=2019-07-07&se=2020-11-18T23%3A31%3A30Z&sr=c&sp=rl&sig=YzLMjEzJP8W3zhGSp5m2JVzphr%2FgNVm0xRO%2FnV4wNJk%3D). It will even add a link to the build artifacts in case you need to download the cross DAC to debug a linux dump on Windows. 

> Go to your build by navigating to [your build](https://dnceng.visualstudio.com/public/_build/results?buildId=870703&view=artifacts&type=publishedArtifacts) and download the `CoreCLRCrossDacArtifacts`, then choose the cross DAC you need depending on the flavor.

stephentoub replied:
I see.  I totally misunderstood the purpose of this documentation. It's not documentation for someone to read in the dotnet/runtime repo.  It's effectively part of our CI tooling. Thanks :)

safern replied:
> I see. I totally misunderstood the purpose of this documentation. It's not documentation for someone to read in the dotnet/runtime repo. It's effectively part of our CI tooling. Thanks :)

Yeah, that's why I put it under `eng\testing`... however, it can serve as documentation as well, but it will just miss getting this piece of information. I could add it and then strip it out when generating the template if you'd like that?

stephentoub replied:
> I could add it and then strip it out when generating the template if you'd like that?

Nah, it's fine, but thanks.

stephentoub commented on eng/testing/debug-dump-template.md:
```diff
@@ -0,0 +1,126 @@
+### Debug with WinDbg
+
+1. Install [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos).
+2. Load the dump with a recent WinDbg version for it to load sos automatically (if not you can run `.update sos`).
```
It's worth noting that the bitness of windbg has to match the bitness of the dump.

safern replied:
Thanks for the suggestion, will include that.

danmosemsft commented:
> However, once we have a symbol support for helix: dotnet/arcade#6480 -- we could very much instead generate a python script on the fly, that people would download and would setup everything for them, basically run the steps in the .md file automatically.

I think this would be valuable and worthwhile. I am interested in all the repos not just ours. What would it take to make this work for all repos -- why is this not in the arcade repo? We are a little special as we do'nt use vstest, but beyond that the dumps are gathered for all repos and presented on this same page -- shouldn't they all see this md?

stephentoub reviewed (commented):
Thanks for putting this together.

hoyosjs commented:
> Another thing -- I am interested in making this easy in all the repos not just ours. What would it take to make this work for all repos -- why is this not in the arcade repo? We are a little special as we do'nt use vstest, but beyond that the dumps are gathered for all repos and presented on this same page -- shouldn't they all see this md?

Partially yes. The runtime has a lot of the symbol behavior that's not so default about the DAC. Once we publish a build that flows to other repos, the dac and coreclr symbols are published. Makes a lot of the setup automatic. At that point the default docs we have for diagnostics should work nicely. We are currently working on cleaning them up all up, so if you have any feedback please let us know so that we can incorporate it and make it as inclusive as possible for all type of experiences and workflows.

danmosemsft commented:
> Partially yes. The runtime has a lot of the symbol behavior that's not so default about the DAC. Once we publish a build that flows to other repos, the dac and coreclr symbols are published. Makes a lot of the setup automatic. At that point the default docs we have for diagnostics should work nicely. We are currently working on cleaning them up all up, so if you have any feedback please let us know so that we can incorporate it and make it as inclusive as possible for all type of experiences and workflows.

OK. Even if in other repos the dump debugging experience will be "less fiddly" I think there is still value in having some doc prominently linked like this next to the dump file, to encourage folks to crack it. I just want to make sure our ultimate goal is all repos.

safern commented:
> Another thing -- I am interested in making this easy in all the repos not just ours. What would it take to make this work for all repos -- why is this not in the arcade repo? We are a little special as we do'nt use vstest, but beyond that the dumps are gathered for all repos and presented on this same page -- shouldn't they all see this md?

That is the end goal. That is what I was trying to tackle at first; but after I synced up with @Chrisboh he suggested doing it here first until the engineering team has a user story on the diagnostics experience. The problem is that aspnet and other repos don't upload symbols as part of their payloads, so to get the matching symbols is a different story (I would expect roslyn will be the same whenever  they move to helix if they do), because it slows down the jobs at setup when downloading payloads. So that is why I opened: https://github.com/dotnet/arcade/issues/6480. 

I was thinking of having some sort of a symbol server correlated to the helix jobs so that we can get the matching symbols from and then all repos would be similar with slight differences, then we can have a template and customize it per repo based on switches or even have an automatic script that setups everything for you which is what @markwilkie is working on. 

I think this is a first step towards getting there and when the time comes we can unify this and make it central in arcade.

safern commented:
Also, cc: @ericstj

safern committed:
PR feedback

safern commented:
This is now ready. I addressed all feedback and also figured out a way to generate a URL to the Cross DAC artifact from the build that the crash happened so people just click that URL from the MD file and get the cross DAC.

This is the resulting MD file:
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-9f01d868602d4fad98/Microsoft.CSharp.Tests/how-to-debug-dump.md?sv=2019-07-07&se=2020-11-24T01%3A13%3A05Z&sr=c&sp=rl&sig=rsyay6SOZRTbcxskTmMWcydBZAIg8VPhT2mEEB0CcAE%3D

![image](https://user-images.githubusercontent.com/22899328/98059991-6302ee80-1dfd-11eb-9a6f-473b21083e16.png)

For CoreCLR runtime tests, I will do the work to generate the MD file in a separate PR.

hoyosjs reviewed (approved):
This looks good as is. Thanks

safern commented:
Failure is: https://github.com/dotnet/runtime/issues/44214

ViktorHofer reviewed (commented):
LGTM, thanks 👍
